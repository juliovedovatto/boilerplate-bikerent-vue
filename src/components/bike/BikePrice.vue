<script lang="ts">
import { defineComponent, type PropType } from 'vue'

export type PropRate = 'daily' | 'weekly' | 'monthly' | 'anually' | 'none'
export enum PropCurrencyCode {
  USD = 'USD',
  CAD = 'CAD',
  EUR = 'EUR',
  BRL = 'BRL'
}

export default defineComponent({
  name: 'BikePrice',
  props: {
    price: {
      type: Number,
      required: true
    },
    currency: {
      type: String as PropType<PropCurrencyCode>,
      required: true
    },
    rate: {
      type: String as PropType<PropRate>,
      default: 'nome'
    }
  },
  computed: {
    formattedPrice() {
      const formatter = new Intl.NumberFormat('en-US', {
        style: 'currency',
        currency: this.currency,
        currencyDisplay: 'symbol',
        maximumFractionDigits: this.price % 1 > 0 ? 2 : 0
      })
      const [{ value: currency }, { value: price }] = formatter.formatToParts(this.price)

      return `${price} ${currency}`
    },
    rateLabel() {
      let label = ''
      switch (this.rate) {
        case 'daily':
          label = 'Day'
          break
        case 'weekly':
          label = 'Week'
          break
        case 'monthly':
          label = 'Month'
          break
        case 'anually':
          label = 'Year'
          break
        default:
      }

      return label
    }
  }
})
</script>

<template>
  <div class="bike__price">
    <strong>{{ formattedPrice }}</strong>
    <template v-if="rateLabel">/ {{ rateLabel }}</template>
  </div>
</template>

<style lang="scss" scoped>
@include block('bike__price') {
  font-weight: 600;
  font-size: 16px;
  line-height: 23px;

  strong {
    font-weight: 800;
    font-size: 24px;
    line-height: 31px;
  }
}
</style>
