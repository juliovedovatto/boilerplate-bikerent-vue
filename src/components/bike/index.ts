export { default as BikeCard } from './BikeCard.vue'
export { default as BikePrice } from './BikePrice.vue'
export { default as BikeBookmark } from './BikeBookmark.vue'
export { default as BikeList } from './BikeList.vue'
