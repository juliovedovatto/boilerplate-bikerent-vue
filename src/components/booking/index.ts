export { default as BookingAddressMap } from './BookingAddressMap.vue'
export { default as BookingPricing } from './BookingPricing.vue'
