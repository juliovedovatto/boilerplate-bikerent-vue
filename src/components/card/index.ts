export { default as BikeCard } from './BikeCard.vue'
export { default as Card } from './Card.vue'
