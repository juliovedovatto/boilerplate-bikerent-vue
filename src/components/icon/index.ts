export { default as Icon, IconType, type IconSize } from './Icon.vue'
