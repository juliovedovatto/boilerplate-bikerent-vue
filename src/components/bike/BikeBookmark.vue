<script lang="ts">
import { Icon, IconType } from '@/components/icon'
import { defineComponent } from 'vue'

export default defineComponent({
  name: 'BikeBookmark',
  components: {
    Icon
  },
  props: {
    active: {
      type: Boolean,
      default: false
    }
  },
  emits: ['update:active'],
  computed: {
    iconType() {
      return this.active ? IconType.SOLID : IconType.REGULAR
    }
  },
  methods: {
    handleToggleActive() {
      this.$emit('update:active', !this.active)
    }
  }
})
</script>

<template>
  <div class="bike-bookmark" @click.prevent="handleToggleActive">
    <icon :type="iconType" size="xl">heart</icon>
  </div>
</template>

<style lang="scss" scoped>
.bike-bookmark {
  position: relative;
  cursor: pointer;
}
</style>
