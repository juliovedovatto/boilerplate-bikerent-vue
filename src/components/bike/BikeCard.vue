<script lang="ts">
import { defineComponent } from 'vue'

import Card from '@/components/card/Card.vue'
import { Chip } from '@/components/chip'
import BikePrice, { PropCurrencyCode } from './BikePrice.vue'
import BikeBookmark from './BikeBookmark.vue'

export default defineComponent({
  name: 'BikeCard',
  components: {
    BikePrice,
    BikeBookmark,
    Card,
    Chip
  },
  data: () => ({
    isBookmarked: false
  }),
  computed: {
    currency() {
      return PropCurrencyCode.EUR
    }
  }
})
</script>

<template>
  <div class="bike-card">
    <card>
      <template #title>
        <div class="bike-card__title">
          <h3><a href="#">Kent Flexer</a></h3>
        </div>
      </template>

      <template #image>
        <div class="bike-card__image">
          <a href="#">
            <img
              src="https://media.istockphoto.com/photos/blue-modern-mens-mid-drive-motor-city-touring-or-trekking-e-bike-picture-id1338461762?s=612x612"
            />
          </a>
        </div>

        <div class="bike-card__bookmark">
          <bike-bookmark v-model:active="isBookmarked" />
        </div>
      </template>

      <template #content>
        <div class="flex items-center">
          <chip>All terrain bike</chip>

          <div class="ml-auto">
            <bike-price :price="25" :currency="currency" rate="daily" />
          </div>
        </div>
      </template>
    </card>
  </div>
</template>

<style lang="scss" scoped>
.bike-card {
  position: relative;
  max-width: 400px;

  @include element('title') {
    @apply .pb-2, .mb-2;

    border-bottom: 1px solid get-theme-color('gray');
  }

  @include element('bookmark') {
    @apply .absolute;

    right: 15px;
    top: 15px;
  }

  @include element('image') {
    @apply .mb-4, .px-6, py-4;

    height: 145px;

    img {
      width: 100%;
      height: 100%;
      object-fit: cover;
      object-position: center;
    }
  }
}
</style>
