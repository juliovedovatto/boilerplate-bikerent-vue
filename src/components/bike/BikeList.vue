<script lang="ts">
/* global Bike */

import BikeCard from './BikeCard.vue'

import { defineComponent, type PropType } from 'vue'

export default defineComponent({
  name: 'BikeList',
  components: {
    BikeCard
  },
  props: {
    items: {
      type: Array as PropType<Bike[]>,
      default: () => []
    }
  },
  data: () => ({})
})
</script>

<template>
  <div class="bike-list grid grid-cols-3 gap-y-6">
    <template v-for="(item, i) in items" :key="`bike-${i}`">
      <bike-card :data="item" />
    </template>
  </div>
</template>

<style lang="scss" scoped>
.bike-list {
  position: relative;
}
</style>
