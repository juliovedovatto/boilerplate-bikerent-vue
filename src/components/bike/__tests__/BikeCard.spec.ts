import { BikeBookmark, BikeCard, BikePrice } from '@/components/bike'
import { describe, expect, it, vi } from 'vitest'

import { CurrencyCode } from '@/core/config'
import { mockIntersectionObserver } from '@/core/helpers/tests'
import { mockedBike } from '@/mocks/'
import { mount } from '@vue/test-utils'
import router from '@/router'

describe('BikeCard', () => {
  vi.stubGlobal('IntersectionObserver', mockIntersectionObserver)

  it('renders correctly', () => {
    const wrapper = mount(BikeCard, { props: { data: mockedBike }, global: { plugins: [router] } })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('displays title', () => {
    const wrapper = mount(BikeCard, { props: { data: mockedBike, fluid: true }, global: { plugins: [router] } })

    expect(wrapper.find('.bike-card__title').text()).equals(mockedBike.name)
  })

  it('renders url', () => {
    const wrapper = mount(BikeCard, { props: { data: mockedBike, fluid: true }, global: { plugins: [router] } })
    const { fullPath: routeStub } = router.resolve({ name: 'details', params: { id: mockedBike.id } })

    expect(wrapper.findComponent({ name: 'router-link' }).exists()).toBe(true)
    expect(wrapper.findComponent({ name: 'router-link' }).find('a').attributes('href')).equals(routeStub)
  })

  it('renders bookmark button', () => {
    const wrapper = mount(BikeCard, { props: { data: mockedBike, fluid: true }, global: { plugins: [router] } })

    expect(wrapper.findComponent(BikeBookmark).exists()).toBe(true)
  })

  it('renders image', () => {
    const wrapper = mount(BikeCard, { props: { data: mockedBike, fluid: true }, global: { plugins: [router] } })

    expect(wrapper.find('.bike-card__image img').exists()).toBe(true)
  })

  it('renders price correctly', () => {
    const wrapper = mount(BikeCard, { props: { data: mockedBike, fluid: true }, global: { plugins: [router] } })
    const bikePriceStub = mount(BikePrice, {
      props: { price: mockedBike.rate, currency: CurrencyCode.EUR, rate: 'daily' }
    })

    expect(wrapper.findComponent(BikePrice).exists()).toBe(true)
    expect(wrapper.findComponent(BikePrice).text()).equal(bikePriceStub.text())
  })

  it('displays fluid layout', () => {
    const wrapper = mount(BikeCard, { props: { data: mockedBike, fluid: true }, global: { plugins: [router] } })

    expect(wrapper.classes()).includes('bike-card--fluid')
  })
})
