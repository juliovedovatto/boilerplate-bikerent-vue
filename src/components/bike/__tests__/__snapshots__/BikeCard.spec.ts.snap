// Vitest Snapshot v1

exports[`BikeCard > renders correctly 1`] = `
"<div class=\\"bike-card\\" data-v-69ad22d1=\\"\\">
  <div class=\\"card\\" data-v-541f45fa=\\"\\" data-v-69ad22d1=\\"\\">
    <div class=\\"card__image\\" data-v-541f45fa=\\"\\">
      <div class=\\"bike-card__image\\" data-v-69ad22d1=\\"\\"><a href=\\"/details/1\\" class=\\"\\" data-v-69ad22d1=\\"\\"><img class=\\"image image--lazy\\" data-v-25f78567=\\"\\" data-v-69ad22d1=\\"\\"></a></div>
      <div class=\\"bike-card__bookmark\\" data-v-69ad22d1=\\"\\">
        <div class=\\"bike-bookmark\\" data-v-3c02c978=\\"\\" data-v-69ad22d1=\\"\\">
          <div class=\\"icon text-base\\" data-v-004c41b7=\\"\\" data-v-3c02c978=\\"\\"><em class=\\"fa-regular fa-heart\\" data-v-004c41b7=\\"\\"></em></div>
        </div>
      </div>
    </div>
    <div class=\\"card__title\\" data-v-541f45fa=\\"\\">
      <div class=\\"bike-card__title\\" data-v-69ad22d1=\\"\\">
        <h3 data-v-69ad22d1=\\"\\"><a href=\\"/details/1\\" class=\\"\\" data-v-69ad22d1=\\"\\">Murazik, Thiel and Robel</a></h3>
      </div>
    </div>
    <div class=\\"card__content\\" data-v-541f45fa=\\"\\">
      <div class=\\"flex items-center\\" data-v-69ad22d1=\\"\\">
        <div class=\\"chip chip--xs chip--secondary\\" data-v-da85bae5=\\"\\" data-v-69ad22d1=\\"\\">
          <div class=\\"chip__content\\" data-v-da85bae5=\\"\\">Cyclocross Bicycle</div>
        </div>
        <div class=\\"ml-auto\\" data-v-69ad22d1=\\"\\">
          <div class=\\"bike__price\\" data-v-0d192dd2=\\"\\" data-v-69ad22d1=\\"\\"><strong data-v-0d192dd2=\\"\\">135 â‚¬</strong>/ Day</div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
